apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,274 @@
+    }
+  });
+
+  var ControlCollectionIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName, keyName) {
+      this.collectionName = collectionName;
+      this.keyName = keyName;
+    },
+    generate: function() {
+      return this.keyName;
+    },
+    validate: function(id) {
+      return (id === this.keyName);
+    }
+  });
+
+  var WebCryptoTransformer = Kinto.createRemoteTransformer({
+    constructor: function(collectionName, fswc) {
+      if (!fswc.bulkKeyBundle) {
+        throw new Error(`Attempt to register Transformer with no bulk key bundl\
+e!`);
+      }
+      this.collectionName = collectionName;
+      this.fswc = fswc;
+    },
+    encode: function(record) {
+      return this.fswc.encrypt(record.payload, this.collectionName)
+      .then(payloadEnc => {
+        record.payload = JSON.stringify(payloadEnc);
+        return record;
+      });
+    },
+    decode: function(record) {
+      // Allowing JSON.parse errors to bubble up to the errors list in the
+      // syncResults:
+      return this.fswc.decrypt(JSON.parse(record.payload), this.collectionName)
+      .then(payloadDec => {
+        record.payload = payloadDec;
+        return record;
+      });
+    }
+  });
+
+  var SyncEngine = function(options) {
+    if (typeof options !== 'object') {
+      throw new Error('options should be an Object');
+    }
+    if (typeof options.URL !== 'string') {
+      throw new Error('options.URL should be a String');
+    }
+    if (typeof options.assertion !== 'string') {
+      throw new Error('options.assertion should be a String');
+    }
+    if (typeof options.xClientState !== 'string') {
+      throw new Error('options.xClientState should be a String');
+    }
+    if (typeof options.kB !== 'string') {
+      throw new Error('options.kB should be a String');
+    }
+
+    this._kB = options.kB;
+    this._collections = {};
+    this._controlCollections = {};
+    this._adapters = {};
+    this._fswc = new FxSyncWebCrypto();
+    this._initKinto({
+       URL: options.URL,
+       assertion: options.assertion,
+       xClientState: options.xClientState
+    });
+  };
+
+  SyncEngine.prototype = {
+    _initKinto: function(kintoCredentials) {
+      this._kinto = new Kinto({
+        bucket: 'syncto',
+        remote: kintoCredentials.URL,
+        headers: {
+          'Authorization': 'BrowserID ' + kintoCredentials.assertion,
+          'X-Client-State': kintoCredentials.xClientState
+        }
+      });
+      var addControlCollection = (collectionName, keyName) => {
+        this._controlCollections[collectionName] =
+            this._kinto.collection(collectionName);
+        this._controlCollections[collectionName].use(
+            new ControlCollectionIdSchema(collectionName, keyName));
+      };
+      addControlCollection('meta', 'global');
+      addControlCollection('crypto', 'keys');
+    },
+
+    _getCollection: function(collectionName, isControlCollection=false) {
```

Comment:
Also, can't we infer if it's a control collection instead of depending on this flag?